### PR TITLE
benchmark: STM benchmark

### PR DESCRIPTION
**Pull request by heyitsanthony**

### CODE REVIEW

**Comment by xiang90** [comment]
can you show an example output result?

**Comment by heyitsanthony** [comment]
```
[anthony@benchmark]$ ./benchmark stm --clients=8 --conns=4 --keys=10 --keys-per-txn=2 --total=1000
1000 / 1000 Boooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo! 100.00 % 1s

Summary:
  Total:	1.5196 secs.
  Slowest:	0.0736 secs.
  Fastest:	0.0039 secs.
  Average:	0.0121 secs.
  Stddev:	0.0086 secs.
  Requests/sec:	658.0514

Response time histogram:
  0.004 [1]	|
  0.011 [575]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.018 [245]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.025 [100]	|∎∎∎∎∎∎
  0.032 [40]	|∎∎
  0.039 [24]	|∎
  0.046 [6]	|
  0.053 [3]	|
  0.060 [3]	|
  0.067 [1]	|
  0.074 [2]	|

Latency distribution:
  10% in 0.0055 secs.
  25% in 0.0063 secs.
  50% in 0.0086 secs.
  75% in 0.0149 secs.
  90% in 0.0229 secs.
  95% in 0.0294 secs.
  99% in 0.0457 secs.
```

File: tools/benchmark/cmd/stm.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	"time"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/cheggaaa/pb"
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/spf13/cobra"
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	v3 "github.com/coreos/etcd/clientv3"
+	v3sync "github.com/coreos/etcd/clientv3/concurrency"
+)
+
+// stmCmd represents the put command
```
Comment (xiang90) [inline tools/benchmark/cmd/stm.go]: the stm command

File: tools/benchmark/cmd/stm.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	stmValSize      int
+	stmWritePercent int
+	mkSTM           func(context.Context, *v3.Client, func(v3sync.STM) error) (*v3.TxnResponse, error)
+)
+
+func init() {
+	RootCmd.AddCommand(stmCmd)
+
+	stmCmd.Flags().StringVar(&stmIsolation, "isolation", "r", "Repeatable Reads (r) or Serializable (s)")
+	stmCmd.Flags().IntVar(&stmKeyCount, "keys", 1, "Number of keys")
```
Comment (xiang90) [inline tools/benchmark/cmd/stm.go]: Total number of keys that stm accesses during benchmark.

File: tools/benchmark/cmd/stm.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	stmCmd.Flags().IntVar(&stmTotal, "total", 10000, "Total number of completed STM transactions")
+	stmCmd.Flags().IntVar(&stmKeysPerTxn, "keys-per-txn", 1, "Number of keys to access per transaction")
+	stmCmd.Flags().IntVar(&stmWritePercent, "txn-wr-percent", 50, "Percentage of keys to overwrite per transaction")
+	stmCmd.Flags().IntVar(&valSize, "val-size", 8, "Value size of put request")
```
Comment (xiang90) [inline tools/benchmark/cmd/stm.go]: value size of the put request in stm.

**Comment by xiang90** [comment]
LGTM

**Comment by xiang90** [comment]
Ideally we should try different keys per txn and change wr percent to see the benchmark result matches our expectation.

**Comment by philips** [comment]
The term "stm" may not be immediately recognizable to folks. Could you add the definition here? https://github.com/coreos/etcd/pull/4802/files#diff-b4d9647cb3da8d6cb5ca7ef7158b9b72R34

**Pull request merged**